nixos/modules/services/networking/firewalld/service.nix

```diff
@@ -0,0 +1,121 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.firewalld;
+  format = pkgs.formats.xml { };
+  lib' = import ./lib.nix { inherit lib; };
+  inherit (lib')
+    filterNullAttrs
+    mkXmlAttr
+    portProtocolOptions
+    toXmlAttrs
+    ;
+  inherit (lib) mkOption;
+  inherit (lib.types)
+    attrsOf
+    listOf
+    nonEmptyStr
+    nullOr
+    strMatching
+    submodule
+    ;
+in
+{
+  options.services.firewalld.services = mkOption {
+    description = ''
+      firewalld service configuration files. See {manpage}`firewalld.service(5)`.
+    '';
+    default = { };
+    type = attrsOf (submodule {
+      options = {
+        version = mkOption {
+          type = nullOr nonEmptyStr;
+          description = "Version of the service.";
+          default = null;
+        };
+        short = mkOption {
+          type = nullOr nonEmptyStr;
+          description = "Short description for the service.";
+          default = null;
+        };
+        description = mkOption {
+          type = nullOr nonEmptyStr;
+          description = "Description for the service.";
+          default = null;
+        };
+        ports = mkOption {
+          type = listOf (submodule portProtocolOptions);
+          description = "Ports of the service.";
+          default = [ ];
+        };
+        protocols = mkOption {
+          type = listOf nonEmptyStr;
+          description = "Protocols for the service.";
+          default = [ ];
+        };
+        sourcePorts = mkOption {
+          type = listOf (submodule portProtocolOptions);
+          description = "Source ports for the service.";
+          default = [ ];
+        };
+        destination = mkOption {
+          type = submodule {
+            options = {
+              ipv4 = mkOption {
+                type = nullOr (strMatching "([0-9]{1,3}\\.){3}[0-9]{1,3}(/[0-9]{1,2})?");
+                description = "IPv4 destination.";
+                default = null;
+              };
+              ipv6 = mkOption {
+                type = nullOr (strMatching "[0-9A-Fa-f:]{3,39}(/[0-9]{1,3})?");
+                description = "IPv6 destination.";
+                default = null;
+              };
+            };
+          };
+          description = "Destinations for the service.";
+          default = { };
+        };
+        includes = mkOption {
+          type = listOf nonEmptyStr;
+          description = "Services to include for the service.";
+          default = [ ];
+        };
+        helpers = mkOption {
+          type = listOf nonEmptyStr;
+          description = "Helpers for the service.";
+          default = [ ];
+        };
+      };
+    });
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment.etc = lib.mapAttrs' (
+      name: value:
+      lib.nameValuePair "firewalld/services/${name}.xml" {
+        source = format.generate "firewalld-service-${name}.xml" {
+          service = filterNullAttrs (
+            lib.mergeAttrsList [
```

Comment:
Note: `mergeAttrsList` is not recursive, it just uses the `//` operator under the hood.

In this case, since you only have two attrsets being merged, I'd recomend using the operator yourself so there are fewer potential points of confusion.

Or alternatively, if you need a recursive update, use [`recursiveUpdate`](https://noogle.dev/f/lib/recursiveUpdate).

It may also make sense to have filtered out the nulls _before_ doing the merge, in case a LHS non-null is shadowed by a RHS null?

Reply:
Also lib.mergeAttrsList is aligned with what we have in zone.nix.